### PR TITLE
Having offset_line_col return the byte offset of columns is annoying to use.

### PR DESCRIPTION
**Pull request by ltratt**
Byte offsets are efficient ways of indexing UTF-8 strings, but are confusing to users, since bytes and characters aren't always the same thing. In other words, in the face of multi-byte characters, offset_line_col gave confusing results.

This commit fixes this, returning the character offset for the column which is nearly always what users want. To make this clear, this commit also renames `offset_line_col` to `line_col`. This is a breaking change, but probably the right thing to do.

[I'm not sure why this didn't occur to me earlier -- probably I made the classic mistake of forgetting that byte and character offsets aren't the same thing in UTF-8. Besides, before the `surrounding_line` function was added, there wasn't really an easy way to perform the necessary calculation.]

### CODE REVIEW

File: lrlex/src/lib/lexer.rs [hunk 1]
```diff
@@ -376,31 +382,53 @@ if 'IF'
...
+
+        let mut lexer = lexerdef.lexer(" a\n❤ b");
+        let lexemes = lexer.all_lexemes().unwrap();
+        assert_eq!(lexemes.len(), 3);
+        assert_eq!(lexer.line_col(lexemes[0].start()), (1, 2));
+        assert_eq!(lexer.line_col(lexemes[1].start()), (2, 1));
+        assert_eq!(lexer.line_col(lexemes[2].start()), (2, 3));
+        assert_eq!(lexer.surrounding_line_str(lexemes[0].start()), " a");
+        assert_eq!(lexer.surrounding_line_str(lexemes[1].start()), "❤ b");
+        assert_eq!(lexer.surrounding_line_str(lexemes[1].start()), "❤ b");
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs]: I think this last assert was meant to be `assert_eq!(lexer.surrounding_line_str(lexemes[2].start()), "b");`

Comment (ltratt) [reply lrlex/src/lib/lexer.rs]: Oops, deliberate mistake! [I might be lying.] I'll fix it.

**Comment by ltratt** [comment]
Fix pushed.

**Comment by ptersilie** [comment]
Rest looks good. Please squash.

File: lrlex/src/lib/lexer.rs [hunk 2]
```diff
@@ -423,7 +423,7 @@ if 'IF'
         assert_eq!(lexer.line_col(lexemes[2].start()), (2, 3));
         assert_eq!(lexer.surrounding_line_str(lexemes[0].start()), " a");
         assert_eq!(lexer.surrounding_line_str(lexemes[1].start()), "❤ b");
-        assert_eq!(lexer.surrounding_line_str(lexemes[1].start()), "❤ b");
+        assert_eq!(lexer.surrounding_line_str(lexemes[2].start()), "❤ b");
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs]: Wait, should the last string only be `"b"`?

Comment (ltratt) [reply lrlex/src/lib/lexer.rs]: No, because it's the whole line that that lexeme is contained in.

Comment (ptersilie) [reply lrlex/src/lib/lexer.rs]: Either I'm missing something, or the tests shouldn't have passed.

Comment (ptersilie) [reply lrlex/src/lib/lexer.rs]: Ah, wait, this is testing `surrounding_line_str`. Nevermind.

Comment (ptersilie) [reply lrlex/src/lib/lexer.rs]: Wouldn't it make more sense to test here if the lexemes are correct? e.g.
```
assert_eq!(lexemes[1], "❤");
assert_eq!(lexemes[2], "b");
```

Comment (ltratt) [reply lrlex/src/lib/lexer.rs]: A unicode test for the pure lexer is a good idea -- I'll add that in.

**Comment by ltratt** [comment]
I've added a multibyte test for the plain lexer. See what you think.

File: lrlex/src/lib/lexer.rs [hunk 3]
```diff
@@ -368,6 +368,31 @@ if 'IF'
...
+
+        let mut lexer = lexerdef.lexer("a ❤ a");
+        let lexemes = lexer.all_lexemes().unwrap();
+        assert_eq!(lexemes.len(), 3);
+        let lex1 = lexemes[0];
+        assert_eq!(lex1.start(), 0);
+        assert_eq!(lex1.len(), Some(1));
+        let lex2 = lexemes[1];
+        assert_eq!(lex2.start(), 2);
+        assert_eq!(lex2.len(), Some(3));
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs]: Would it be difficult to add another check here, that checks the content of the lexeme, just so we also test, that the lexer doesn't lex something like ` \xe2\x9d` as the second lexeme (e.g. space + the first two characters of ❤)?

Comment (ltratt) [reply lrlex/src/lib/lexer.rs]: Done -- try the newest commit.

Comment (ptersilie) [reply lrlex/src/lib/lexer.rs]: Perfect!

**Comment by ptersilie** [comment]
Looking good now. Please squash!

**Comment by ltratt** [comment]
Squashed (note that I kept the testing-the-lexer commit separate, because it's really stand alone).

**Pull request merged**